### ISSUE TITLE
Single path optimisation for Collection w/ offsets broken

### ISSUE THREAD

**Issue by dstansby** [matplotlib/matplotlib#16496]
This is the root cause of #16461. The single path optimisation does not correctly take into account offsets:

https://github.com/matplotlib/matplotlib/blob/acc7ecbfbe5c6e944a45fa08cc39142098149893/lib/matplotlib/collections.py#L354-L362

As an example, this code creates two collections which differ only by their `edgecolors`, and have an offset of `[0.5, 0.5]`. In both cases the collection should be drawn at the offset, but only the collection with `edgecolors=None` is correct, because this does not follow the 'optimised' code path above.

```python
import matplotlib.pyplot as plt
import matplotlib.collections as mcoll
import matplotlib.transforms as mtransforms
import numpy as np

polygon = np.array([[-1, -1], [-1, 1], [1, 1], [1, -1]]) * 0.1
offsets = np.array([0.5, 0.5])

fig, axs = plt.subplots(ncols=2)
for ax, edgecolors in zip(axs, [None, 'face']):
    collection = mcoll.PolyCollection(
        [polygon],
        edgecolors=edgecolors,
        linewidths=[1],
        offsets=offsets,
        transOffset=mtransforms.IdentityTransform(),
        offset_position="data"
        )
    ax.add_collection(collection)

plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/6197628/74437173-9c569100-4e5f-11ea-939c-ac668b0ddc05.png)

**Comment by anntzer** [comment]
FWIW I think offset_position="data" should be deprecated.  See https://github.com/matplotlib/matplotlib/pull/13696, https://github.com/matplotlib/matplotlib/issues/16461#issuecomment-586181712.

**Comment by QuLogic** [comment]
Changes to `hexbin` aside, possibly related to #15946?

**Comment by anntzer** [comment]
Not directly, I think?  At least the issue here seems to be speficically about offset_position="data" and is fixed by https://github.com/matplotlib/matplotlib/pull/13696, but #15946 doesn't use offset_position="data" and appears unaffected by my pr.

**Comment by QuLogic** [comment]
Ah, I didn't realize `offset_position` was a collection thing.

**Comment by tacaswell** [comment]
@dstansby RC for which version?

**Comment by tacaswell** [comment]
ok, the fixes here seem to be:

 - the full deprecation of offset_position='data' (aka #16399 )
 - in the case of single path optimization and data add transData to transOffset before we pass it down to `draw_marker`
 - detect the data case and set do_single_path_optimization to False

in order of most correct to simplest.  If we want to fix this for 3.2, we should go with the last one, if we want to target this at 3.3 we should go with the first one.

**Comment by anntzer** [comment]
The problem of the last approach is that you'd need to change the baseline images, as discussed above.
Again, I think we should just postpone this to 3.3, this bug has likely been present for a long time.

**Comment by tacaswell** [comment]
I missed #16517 earlier (which basically implements the last one).

Sorry, I am being very dense, why does that require regenerating test images?

**Comment by anntzer** [comment]
Becauses images with and without single path optimization are different (because single path optimization is based on stamping which means snapping to nearest pixels, whereas without SPO, you don't snap).

**Comment by anntzer** [comment]
Ah but wait, this was resolved via https://github.com/matplotlib/matplotlib/pull/16517#issuecomment-586644187.

**Comment by QuLogic** [comment]
This is as old as 2.0.2 at least. In 1.5.3, _both_ paths are drawn at (0, 0), which I guess is a slight improvement. But this is a very long-standing bug.

**Comment by goretkin** [comment]
`Axes.scatter` produces a `PathCollection` with `offset_position='screen'`, but with `offsets` given in data coordinates, as far as I can see. So fixing this would break `Axes.scatter`, unless it was also fixed.

**Comment by QuLogic** [comment]
#13696 deprecated `offset_position='data'`, so replacing:

```python
        transOffset=mtransforms.IdentityTransform(),
        offset_position="data"
```
with
```python
        transOffset=mtransforms.AffineDeltaTransform(ax.transData),
```
works.

I think we can close this now, as the code path will eventually get deleted?

**Comment by QuLogic** [comment]
This is fixed for 3.3, and it doesn't look like anyone's going to open a simpler PR for 3.2.2.
